keps/sig-release/20190316-rebase-images-to-distroless.md

```diff
@@ -0,0 +1,149 @@
+
+* Add Alpine iptable as base image for kube-proxy. Previously kube-proxy is based on debian iptable image.  (This direction is scrapped, see [issue/39696](https://github.com/kubernetes/kubernetes/issues/39696) for details)
+Rebase busybox images to debian-base.
+* Rebase busybox images to debian-base.
+* Rebase certain alpine images to debian-base
+* Image related works on migrating apiserver-builder to kube-builder
+
+The distroless solution is filed separately in [issue/70249](https://github.com/kubernetes/kubernetes/issues/70249). The doc, as a more up-to-date version, is slightly different than the original issue.
+
+## Proposal
+
+The approaches to rebase different containers vary significantly due to the technical implementations, development plans or even the legacy reasons. For example, [ingress-gce/fuzzer](https://github.com/kubernetes/ingress-gce/blob/64eee7e3521680057b071d5e9bebaa215086a4bc/Dockerfile.fuzzer) was based on alpine and can't be switched to distroless directly due to the fact that it needs shell to redirect the glog file. To allow the images to be based on distroless (which doesn't contain shell), we firstly need to remove the dependency on the shell (use klog instead of glog), and then rebase the image. (Related PR [pr/682](https://github.com/kubernetes/ingress-gce/pull/682]), [pr/666](https://github.com/kubernetes/ingress-gce/pull/666))
+There are 4 main efforts on this KEP.
+
+1. The images are built via bazel. In such case, we will update the bazel BUILD rule to switch to base from distroless image. This method applies for the core containers like kube-apiserver, kube-controller-manager, cloud-controller-manager, kube-scheduler. As for kube-proxy, it is based on debian-iptable which distroless doesn't support iptables yet. Thus, kube-proxy won't be changed.
+2. The images have dependencies that are not supported by distroless. One typical example is the usage of shell. Previously, shell is widely used for redirecting glog output to a certain directory. This use case is no longer needed since we've switched from glog to klog which can accept a flag to specify the log output path. A generic approach is: Remove the dependencies that distroless doesn't support and then rebase the images to distroless (e.g. [issue/1787](https://github.com/kubernetes/autoscaler/issues/1787). Meanwhile, we limit the introduction of new dependencies. (e.g. [pr/74690](https://github.com/kubernetes/kubernetes/pull/74690#discussion_r266037189))
+3. Images based "FROM scratch" is safe to switch to distroless directly.
+4. Images from kubernetes incubator won't be changed directly by this KEP and the release plan is not estimated here. We notify the project OWNERs and we defer to the OWNERs on whether or not those images should be updated.
```

Comment:
Are containers from kubernetes incubator published under k8s.gcr.io? If so, maybe we should consider creating a separate repository for those, especially if they're not going to have the same security requirements.

Reply:
I don't know, but I guess it's the incubator repo owners or release engineers' choice? qq: what's the benefit of separating incubator images from the gcr.k8s.io besides making it clear that they have different security requirements?

I'm a little bit worried this may cause unexpected problems and overheads especially that the container registry is public and anyone (not tracked by us or even not tightly involved in the k8s community) can use it. What do you think?